addons/binding/org.openhab.binding.russound/src/main/java/org/openhab/binding/russound/internal/discovery/RioSystemDeviceDiscoveryService.java

```diff
@@ -0,0 +1,270 @@
+    public void deactivate() {
+        if (_session != null) {
+            try {
+                _session.disconnect();
+            } catch (IOException e) {
+                // ignore
+            }
+            _session.removeListener(_listener);
+            _session = null;
+            _listener = null;
+        }
+
+        // sysHandler.unregisterDiscoveryService();
+    }
+
+    /**
+     * Overridden to do nothing - {@link #scanDevice()} is called by {@link RioSystemHandler} instead
+     */
+    @Override
+    protected void startScan() {
+        // do nothing - started by RioSystemHandler
+
+    }
+
+    /**
+     * Starts a device scan. This will connect to the device and discover the controllers/sources/zones attached to the
+     * device and then disconnect via {@link #deactivate()}
+     */
+    public void scanDevice() {
+        try {
+            final String ipAddress = sysHandler.getRioConfig().getIpAddress();
+            _session = new SocketChannelSession(ipAddress, RioConstants.RioPort);
+            _listener = new WaitingSessionListener();
+            _session.addListener(_listener);
+
+            try {
+                logger.debug("Starting scan of RIO device at " + ipAddress);
+                _session.connect();
+                discoverControllers();
+                discoverSources();
+            } catch (IOException e) {
+                logger.warn("Trying to scan device but couldn't connect: {}", e.getMessage(), e);
+            }
+        } finally {
+            deactivate();
+        }
+    }
+
+    /**
+     * Helper method to discover controllers - this will iterate through all possible controllers (6 of them )and see if
+     * any respond to the "type" command. If they do, we initiate a {@link #thingDiscovered(DiscoveryResult)} for the
+     * controller and then scan the controller for zones via {@link #discoverZones(ThingUID, int)}
+     */
+    private void discoverControllers() {
+        for (int c = 1; c < 7; c++) {
+            try {
+                final String type = sendAndGet("GET C[" + c + "].type", RSP_CONTROLLERNOTIFICATION, 3);
+                if (StringUtils.isNotEmpty(type)) {
+                    logger.debug("Controller #{} found - {}", c, type);
+
+                    final ThingUID thingUID = new ThingUID(RioConstants.BRIDGE_TYPE_CONTROLLER,
+                            sysHandler.getThing().getUID(), String.valueOf(c));
+
+                    final DiscoveryResult discoveryResult = DiscoveryResultBuilder.create(thingUID)
+                            .withProperty(RioControllerConfig.Controller, c).withBridge(sysHandler.getThing().getUID())
+                            .withLabel("Controller #" + c).build();
+                    thingDiscovered(discoveryResult);
+
+                    discoverZones(thingUID, c);
+                }
+            } catch (Exception e) {
+                logger.debug("Exception occurred scanning controller #{} : {}", c, e.getMessage(), e);
```

Comment:
Catching of these exceptions were for my iterative dev process and I forgot to remove them.  They will be gone on the commit